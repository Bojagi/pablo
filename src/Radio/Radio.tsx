import React, { forwardRef } from 'react';
import { css } from '@emotion/react';
import styled from '@emotion/styled';
import { getComponentStyle, transitionTransformer } from '../styleHelpers/getComponentStyle';

import {
  BaseCheckable,
  BaseCheckableProps,
  CheckableBoxProps,
  CheckableHandleProps,
} from '../shared/BaseCheckable';
import { getCustomStyles } from '../utils/useCustomStyles';
import { baseStyle } from '../shared/baseStyle';

const radioBoxSize = (props) => css`
  calc(${getComponentStyle('radio.handleSize.{size}')(props)} + 2 * (${getComponentStyle(
    'radio.innerPadding.{size}'
  )(props)} + ${getComponentStyle('radio.borderWidth')(props)}px))
`;

const RadioBox = styled.div<CheckableBoxProps>`
  ${baseStyle}
  position: relative;
  width: ${radioBoxSize};
  height: ${radioBoxSize};
  border-radius: 50%;
  padding: ${getComponentStyle('radio.innerPadding.{size}')};
  background-color: ${getComponentStyle('radio.backgroundColor')};
  border: ${getComponentStyle('radio.borderWidth')}px solid
    ${getComponentStyle('radio.borderColor')};
  opacity: ${(props) => (props.disabled ? 0.3 : 1)};
  transition: ${getComponentStyle('radio.boxTransition', transitionTransformer)};

  ${(props) =>
    props.focus &&
    css`
      box-shadow: 0 0 0 ${getComponentStyle('radio.focus.outlineSize')(props)}
        ${getComponentStyle('radio.focus.outlineColor')(props)};
    `};
  }
  ${getCustomStyles('radio.styles', 'box')}
`;

const RadioHandle = styled.div<CheckableHandleProps>`
  ${baseStyle}
  width: ${getComponentStyle('radio.handleSize.{size}')};
  height: ${getComponentStyle('radio.handleSize.{size}')};
  transform: scale(${(props: any) => (props.checked ? 1 : 0)});
  border-radius: 50%;
  transition: ${getComponentStyle('radio.handleTransition', transitionTransformer)};
  background-color: ${getComponentStyle('radio.handleColor')};
  ${getCustomStyles('radio.styles', 'handle')}
`;

export interface RadioProps extends BaseCheckableProps {
  value: string;
}

export const Radio = forwardRef<HTMLDivElement, RadioProps>((props: RadioProps, ref) => (
  <BaseCheckable
    ref={ref}
    role="radio"
    componentName="radio"
    componentType="radio"
    componentBox={RadioBox}
    componentHandle={RadioHandle}
    {...props}
  />
));
