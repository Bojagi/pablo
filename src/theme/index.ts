export * from './defaultTheme';
export * from './context';
export * from './PabloThemeProvider';
export * from './useComponentStyle';
