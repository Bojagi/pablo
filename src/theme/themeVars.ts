import isObject from 'isobject';
import { breakpointVars } from './breakpoints';
import { colorVars } from './colors';
import { typographyVars } from './typography';

export const themeVarNames = {
  typography: typographyVars,
  colors: colorVars,
  breakpoints: breakpointVars,
};

function createThemeReferences<T extends Record<string, any>>(obj: T): T {
  return Object.entries(obj).reduce(
    (acc, [key, value]) => ({
      ...acc,
      [key]: isObject(value) ? createThemeReferences(value as T) : `var(--${value})`,
    }),
    {} as any
  );
}

export const themeVars = createThemeReferences(themeVarNames);
