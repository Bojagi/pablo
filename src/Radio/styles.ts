import { getColor } from '../styleHelpers/getColor';
import { getSpacing } from '../styleHelpers/getSpacing';
import { Style } from '../theme/types';
import { checkableBaseStyles, CheckableBaseStyles } from '../shared/checkableBaseStyles';

export interface RadioStyles extends CheckableBaseStyles {
  handleColor: Style;
  groupItemGap: {
    medium: Style;
    small: Style;
  };
}

export const radioStyles: RadioStyles = {
  ...checkableBaseStyles,
  handleColor: getColor('brand'),
  groupItemGap: {
    medium: getSpacing(5),
    small: getSpacing(4),
  },
};
