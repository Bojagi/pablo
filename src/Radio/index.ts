import './styles';

export * from './Radio';
export * from './RadioGroup';
