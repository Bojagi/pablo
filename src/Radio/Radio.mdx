---
name: Radio
route: /radio
menu: Components
---

import { Props, Playground } from 'docz'
import { Box } from '../Box/Box.tsx'
import { Radio } from './Radio.tsx'
import { RadioGroup } from './RadioGroup.tsx'
import { css } from 'styled-components';
import { getSpacing } from '../styleHelpers/index.ts'

# Radio

Form element that allows users to select a single item from a set of options. 
The `Radio` component shares an interface with the `Checkbox` and `Switch` components.

## Examples

### Single radio component

<Playground>
  {() => {
    const [checked, setChecked] = React.useState(false);
    return (
      <Radio checked={checked} onChange={() => setChecked(!checked)}  />
    );
  }}
</Playground>

### Radio selectors with labels

You can pass an optional `label` prop with a `ReactNode` value to show a label next to the checkbox.

<Playground>
  {() => {
    const [selectedDish, setDish] = React.useState(undefined);
    return (
      <Box display="flex" mx={-2}>
        <Radio mx={2} label="Spaghetti" checked={selectedDish === 'spaghetti'} onChange={() => setDish('spaghetti')} />
        <Radio mx={2} label="Pizza" checked={selectedDish === 'pizza'} onChange={() => setDish('pizza')} />
      </Box>
    );
  }}
</Playground>

### Grouped radio selectors

In the above example the selection management is done manually, it makes more sense though to use the `RadioGroup` to
group the values together. The `RadioGroup` component takes the selected `value` and `onChange` callback props.
For this to work the `Radio` components need to be passed a `value` where they should be selected. The `checked` prop is not needed.

<Playground>
  {() => {
    const [selectedDish, setDish] = React.useState(undefined);
    return (
      <RadioGroup value={selectedDish} onChange={setDish}>
        <Radio mx={2} label="Spaghetti" value="spaghetti" />
        <Radio mx={2} label="Pizza" value="pizza" />
      </RadioGroup>
    );
  }}
</Playground>

### Small radio selectors

You can pass the `size` prop with the `small` value which renders a smaller version.

<Playground>
  {() => {
    const [checked, setChecked] = React.useState(false);
    return (
      <Radio size="small" checked={checked} onChange={() => setChecked(!checked)} />
    );
  }}
</Playground>

When you want to set the size on a `RadioGroup` you have to set it in the `RadioGroup` component

<Playground>
  {() => {
    const [selectedDish, setDish] = React.useState(undefined);
    return (
      <RadioGroup size="small" value={selectedDish} onChange={setDish}>
        <Radio mx={2} label="Spaghetti" value="spaghetti" />
        <Radio mx={2} label="Pizza" value="pizza" />
      </RadioGroup>
    );
  }}
</Playground>

### Custom styling

You can overwrite the styles of parts of this component as well. 

<Playground>
  {() => {
    const [checked, setChecked] = React.useState(false);
    const customStyles = {
      box: css`
        background-color: blue;
        order: 2;
      `,
      handle: css`
        background-color: red;
      `,
      label: css`
        text-decoration: underline;
        order: 1;
        margin-right: ${getSpacing(4)};
      `,
    };
    return (
      <Radio
        checked={checked}
        label="weird"
        onChange={() => setChecked(!checked)}
        customStyles={customStyles} 
      />
    );
  }}
</Playground>

## Import

```js
import { Radio, RadioGroup } from '@bojagi/pablo/Radio';
```

## Props
<Props of={Radio} />
