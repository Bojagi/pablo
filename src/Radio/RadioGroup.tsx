import React from 'react';
import styled from 'styled-components';
import { Box, BoxProps } from '../Box';
import { guaranteeArray } from '../utils/guaranteeArray';
import { getComponentStyle } from '../utils/styleHelpers/getComponentStyle';
import { RadioProps, RadioSize } from './Radio';

const RadioGroupItem = styled.div<{ size: RadioSize }>`
  margin-bottom: ${getComponentStyle('radio.groupItemGap.{size}')};
`;

export interface RadioGroupProps extends BoxProps {
  value?: string;
  onChange: (value: string) => void;
  children: React.ReactElement<RadioProps> | React.ReactElement<RadioProps>[];
  size?: RadioSize;
  name?: string;
}

export function RadioGroup({
  value,
  onChange,
  name,
  children,
  size = 'medium',
  ...props
}: RadioGroupProps) {
  return (
    <Box role="radiogroup" {...props}>
      {guaranteeArray(children).map((child) => (
        <RadioGroupItem key={child.props.value} size={size}>
          {React.cloneElement(child, {
            checked: value === child.props.value,
            size,
            name,
            onChange: () => {
              onChange(child.props.value);
            },
          })}
        </RadioGroupItem>
      ))}
    </Box>
  );
}
