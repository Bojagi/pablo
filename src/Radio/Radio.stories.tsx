import React from 'react';
import { Box } from '../Box';
import { Radio, RadioProps } from './Radio';
import { RadioGroup } from './RadioGroup';

export default {
  title: 'Radio',
};

const ControlledRadio = ({ checked: checkedInitial, ...props }: RadioProps) => {
  const [checked, setChecked] = React.useState(checkedInitial);
  return (
    <Radio checked={checked} onChange={() => {}} onClick={() => setChecked(!checked)} {...props} />
  );
};

const baseStory = (args) => (
  <Box flex mx={-1} mb={1}>
    <ControlledRadio mx={1} value="1" name="hans" checked={true} {...args} />
    <ControlledRadio mx={1} value="2" name="hans" checked={false} {...args} />
    <ControlledRadio mx={1} value="3" name="hans" checked={true} disabled={true} {...args} />
  </Box>
);

export const Medium = baseStory.bind(null);
Medium.args = {
  size: 'medium',
};

export const Small = baseStory.bind(null);
Small.args = {
  size: 'small',
};

const BaseRadioGroupStory = (args) => {
  const [value, setValue] = React.useState<string>();
  return (
    <RadioGroup value={value} onChange={setValue} {...args}>
      <Radio value="hans" label="hans" />
      <Radio value="wurst" label="wurst" />
    </RadioGroup>
  );
};

export const RadioGroupMedium = BaseRadioGroupStory.bind(null);
RadioGroupMedium.args = {
  size: 'medium',
};

export const RadioGroupSmall = BaseRadioGroupStory.bind(null);
RadioGroupSmall.args = {
  size: 'small',
};
