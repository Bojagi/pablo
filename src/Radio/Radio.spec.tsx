import { act, cleanup, fireEvent, render } from '@testing-library/react';
import React from 'react';
import { PabloThemeProvider } from '../theme';
import { themeVars } from '../theme/themeVars';
import { Radio } from './Radio';

let onChangeMock;

beforeEach(() => {
  onChangeMock = jest.fn();
});

afterEach(cleanup);

test('Render radio unchecked', () => {
  const { container, getByTestId } = renderComponent({ checked: false });
  expect(container).toMatchSnapshot();
  expect(getByTestId('pbl-radio-handle')).toHaveStyleRule('transform', 'scale(0)');
  expect(getByTestId('pbl-radio-input')).not.toBeChecked();
});

test('Render radio checked', () => {
  const { container, getByTestId } = renderComponent({ checked: true });
  expect(container).toMatchSnapshot();
  expect(getByTestId('pbl-radio-handle')).toHaveStyleRule('transform', 'scale(1)');
  expect(getByTestId('pbl-radio-input')).toBeChecked();
});

test('Render radio with label', () => {
  const { getByTestId } = renderComponent({
    checked: true,
    label: (
      <>
        I am a <strong>label</strong>
      </>
    ),
  });
  expect(getByTestId('pbl-radio-label')).toBeDefined();

  // Label and input are connected via "id"
  expect(getByTestId('pbl-radio-label')).toHaveAttribute('for', 'custom-id');
  expect(getByTestId('pbl-radio-input')).toHaveAttribute('id', 'custom-id');
  expect(getByTestId('pbl-radio-label')).toMatchSnapshot();
  expect(getByTestId('pbl-radio-label')).toHaveTextContent('I am a label');
});

test('Click on radio to trigger onChange event', () => {
  const { getByTestId } = renderComponent({
    checked: false,
    value: 'radio',
    label: 'i am a label',
  });
  expect(onChangeMock).toHaveBeenCalledTimes(0);
  expect(getByTestId('pbl-radio-input')).toBeEnabled();

  act(() => {
    fireEvent.click(getByTestId('pbl-radio-input'), { target: { value: 'radio' } });
  });
  expect(onChangeMock).toHaveBeenCalledTimes(1);
  expect(onChangeMock).toHaveBeenCalledWith('radio', expect.objectContaining({ type: 'change' }));
});

test('Disabled prop should disable input', () => {
  const { getByTestId } = renderComponent({
    checked: true,
    disabled: true,
  });
  expect(getByTestId('pbl-radio-input')).toBeDisabled();
});

test('Render medium radio', () => {
  const { getByTestId } = renderComponent({
    size: 'medium',
    label: 'medium label',
  });
  expect(getByTestId('pbl-radio')).toHaveStyleRule('width', 'calc(12px + 2 * (6px + 1px))');
  expect(getByTestId('pbl-radio')).toHaveStyleRule('height', 'calc(12px + 2 * (6px + 1px))');
  expect(getByTestId('pbl-radio')).toHaveStyleRule('padding', '6px');
  expect(getByTestId('pbl-radio-handle')).toHaveStyleRule('width', '12px');
  expect(getByTestId('pbl-radio-handle')).toHaveStyleRule('height', '12px');
  // Uses subtitle typography
  expect(getByTestId('pbl-radio-label')).toHaveStyleRule(
    'font-size',
    themeVars.typography.subtitle.fontSize
  );
});

test('Render small radio', () => {
  const { getByTestId } = renderComponent({
    size: 'small',
    label: 'small label',
  });
  expect(getByTestId('pbl-radio')).toHaveStyleRule('width', 'calc(8px + 2 * (4px + 1px))');
  expect(getByTestId('pbl-radio')).toHaveStyleRule('height', 'calc(8px + 2 * (4px + 1px))');
  expect(getByTestId('pbl-radio')).toHaveStyleRule('padding', '4px');
  expect(getByTestId('pbl-radio-handle')).toHaveStyleRule('width', '8px');
  expect(getByTestId('pbl-radio-handle')).toHaveStyleRule('height', '8px');
  // Uses paragraph typography
  expect(getByTestId('pbl-radio-label')).toHaveStyleRule(
    'font-size',
    themeVars.typography.paragraph.fontSize
  );
});

test('Use generated id when no "id" prop is set', () => {
  const { getByTestId } = renderComponent({
    id: undefined,
  });

  expect(getByTestId('pbl-radio-input')).toHaveAttribute('id', expect.stringMatching(/:.*?:/));
});

test('Render without onChange function ', () => {
  renderComponent({
    onChange: undefined,
    checked: undefined,
  });
});

test('Forward onFocus and onBlur to input and show focus outline', () => {
  const onFocusMock = jest.fn();
  const onBlurMock = jest.fn();
  const { getByTestId } = renderComponent({
    onChange: undefined,
    checked: undefined,
    onFocus: onFocusMock,
    onBlur: onBlurMock,
  });

  expect(onFocusMock).toHaveBeenCalledTimes(0);
  act(() => {
    fireEvent.focus(getByTestId('pbl-radio-input'));
  });
  expect(onFocusMock).toHaveBeenCalledTimes(1);

  expect(getByTestId('pbl-radio')).toHaveStyleRule(
    'box-shadow',
    `0 0 0 3px ${themeVars.colors.brand.lightest}`
  );

  expect(onBlurMock).toHaveBeenCalledTimes(0);
  act(() => {
    fireEvent.blur(getByTestId('pbl-radio-input'));
  });
  expect(onBlurMock).toHaveBeenCalledTimes(1);

  expect(getByTestId('pbl-radio')).toHaveStyleRule('box-shadow', undefined);
});

function renderComponent(props) {
  return render(
    <PabloThemeProvider>
      <Radio id="custom-id" onChange={onChangeMock} {...props}>
        Hello
      </Radio>
    </PabloThemeProvider>
  );
}
