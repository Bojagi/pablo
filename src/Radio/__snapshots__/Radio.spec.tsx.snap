// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render radio checked 1`] = `
.emotion-0 {
  box-sizing: border-box;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.emotion-1 {
  box-sizing: border-box;
  position: relative;
  width: calc(0.75rem + 2 * (0.375rem + 1px));
  height: calc(0.75rem + 2 * (0.375rem + 1px));
  border-radius: 50%;
  padding: 0.375rem;
  background-color: var(--pbl-theme-colors-common-white);
  border: 1px solid var(--pbl-theme-colors-borders-main);
  opacity: 1;
  -webkit-transition: box-shadow 0.3s ease-in-out;
  transition: box-shadow 0.3s ease-in-out;
}

.emotion-2 {
  box-sizing: border-box;
  width: 0.75rem;
  height: 0.75rem;
  -webkit-transform: scale(1);
  -moz-transform: scale(1);
  -ms-transform: scale(1);
  transform: scale(1);
  border-radius: 50%;
  -webkit-transition: -webkit-transform 0.15s cubic-bezier(0.34, 1.56, 0.64, 1);
  transition: transform 0.15s cubic-bezier(0.34, 1.56, 0.64, 1);
  background-color: var(--pbl-theme-colors-brand-main);
}

.emotion-3 {
  box-sizing: border-box;
  position: absolute;
  display: block;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
  width: 100%;
  height: 100%;
  opacity: 0;
  cursor: pointer;
}

<div>
  <div
    class="emotion-0"
    role="radio"
  >
    <div
      class="emotion-1"
      data-testid="pbl-radio"
    >
      <div
        class="emotion-2"
        data-testid="pbl-radio-handle"
      />
      <input
        checked=""
        class="emotion-3"
        data-testid="pbl-radio-input"
        id="custom-id"
        type="radio"
        value=""
      />
    </div>
  </div>
</div>
`;

exports[`Render radio unchecked 1`] = `
.emotion-0 {
  box-sizing: border-box;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.emotion-1 {
  box-sizing: border-box;
  position: relative;
  width: calc(0.75rem + 2 * (0.375rem + 1px));
  height: calc(0.75rem + 2 * (0.375rem + 1px));
  border-radius: 50%;
  padding: 0.375rem;
  background-color: var(--pbl-theme-colors-common-white);
  border: 1px solid var(--pbl-theme-colors-borders-main);
  opacity: 1;
  -webkit-transition: box-shadow 0.3s ease-in-out;
  transition: box-shadow 0.3s ease-in-out;
}

.emotion-2 {
  box-sizing: border-box;
  width: 0.75rem;
  height: 0.75rem;
  -webkit-transform: scale(0);
  -moz-transform: scale(0);
  -ms-transform: scale(0);
  transform: scale(0);
  border-radius: 50%;
  -webkit-transition: -webkit-transform 0.15s cubic-bezier(0.34, 1.56, 0.64, 1);
  transition: transform 0.15s cubic-bezier(0.34, 1.56, 0.64, 1);
  background-color: var(--pbl-theme-colors-brand-main);
}

.emotion-3 {
  box-sizing: border-box;
  position: absolute;
  display: block;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
  width: 100%;
  height: 100%;
  opacity: 0;
  cursor: pointer;
}

<div>
  <div
    class="emotion-0"
    role="radio"
  >
    <div
      class="emotion-1"
      data-testid="pbl-radio"
    >
      <div
        class="emotion-2"
        data-testid="pbl-radio-handle"
      />
      <input
        class="emotion-3"
        data-testid="pbl-radio-input"
        id="custom-id"
        type="radio"
        value=""
      />
    </div>
  </div>
</div>
`;

exports[`Render radio with label 1`] = `
.emotion-0 {
  box-sizing: border-box;
  font-family: var(--pbl-theme-typography-base-fontFamily);
  font-style: normal;
  font-weight: var(--pbl-theme-typography-base-fontWeight);
  margin: 0;
  font-size: var(--pbl-theme-typography-subtitle-fontSize);
  line-height: var(--pbl-theme-typography-subtitle-lineHeight);
  margin-bottom: var(--pbl-theme-typography-subtitle-marginBottom);
  margin-left: 0.5rem;
}

<label
  class="emotion-0"
  data-testid="pbl-radio-label"
  for="custom-id"
>
  I am a 
  <strong>
    label
  </strong>
</label>
`;
